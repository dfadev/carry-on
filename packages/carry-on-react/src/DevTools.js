import { Component } from "react";
import { register, devTools } from "carry-on-store";

let hasRegistered = false;

export default class DevTools extends Component {
  constructor(props) {
    super(props);
    if (hasRegistered) return;
    hasRegistered = true;
    register(devTools(), props.store);
  }

  render() {
    return this.props.children || null;
  }
}
