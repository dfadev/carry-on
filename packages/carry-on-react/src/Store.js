/** @format **/
import React, { Component } from "react";
import { isFunction } from "./utils";

// create an object with Store and withState components
export default function makeStoreComponents({
  useStore,
  deleteStore,
  connect
}) {
  // a Component that manages state
  class Store extends Component {
    state = connect({
      ...this.props,
      publish: nextState => this.setState(nextState),
      Context: React.createContext({})
    });

    // unmount failsafe to prevent setState calls after unmounting
    componentWillUnmount = () => deleteStore(this.props.id);

    // render the Provider and it's children
    render = () => {
      const { Provider } = useStore(this.props.id).Context;
      return <Provider value={this.state}>{this.props.children}</Provider>;
    };
  }

  // a HOC that wraps the Store component
  const withStore = (opts = {}) => WrappedComponent => {
    const WithStore = props => (
      <Store
        id={opts.id}
        init={isFunction(opts.init) ? opts.init(props) : opts.init}
        producer={opts.producer}
        plugins={opts.plugins}
      >
        <WrappedComponent {...props} />
      </Store>
    );
    return WithStore;
  };

  return { Store, withStore };
}
