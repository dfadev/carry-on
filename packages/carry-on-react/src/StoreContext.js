import { createContext } from "react";

const context = createContext();
context.displayName = "Store";

export default context;
