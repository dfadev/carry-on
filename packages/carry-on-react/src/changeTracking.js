/** @format **/
import { getIn, mutateSet } from "carry-on-utils";

export function compareChanges(changes, affected) {
  const queue = [];
  queue.push({ changes, affected });

  while (queue.length > 0) {
    const item = queue.pop();
    const entries = item.changes;

    for (let i = 0, len = entries.length; i < len; i++) {
      const entry = entries[i];
      const key = entry.key;
      const affectedValue = item.affected[key];

      if (affectedValue === true) return true;
      if (affectedValue !== undefined) {
        const nextChanges = entry.changes;

        if (nextChanges === true) return true;
        queue.push({
          changes: nextChanges,
          affected: affectedValue
        });
      }
    }
  }

  return false;
}

export function createAffectedKeysIndex(path, affected) {
  const affectedStateKeys = {};
  mutateSet(affectedStateKeys, path, {});
  const prefix = getIn(affectedStateKeys, path);

  for (let i = 0; i < affected.length; i++)
    mutateSet(prefix, affected[i], true);

  return affectedStateKeys;
}
