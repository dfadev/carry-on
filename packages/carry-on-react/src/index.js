export { default as State, Data } from "./State";
export { default as withState } from "./withState";
export { default as carryOn } from "./carryOn";
export { default as DevTools } from "./DevTools";
export { default as useCarryOn } from "./useCarryOn";
