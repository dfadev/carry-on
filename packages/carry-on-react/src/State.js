/* eslint-disable react/jsx-props-no-spreading */
import React, { Component } from "react";
import hoistNonReactStatic from "hoist-non-react-statics";
import {
  debounce,
  getIn,
  logger,
  shallowEqual,
  throttle
} from "carry-on-utils";
import { register, connect, subscribe, watchGet } from "carry-on-store";
import ReactDOM from "react-dom";
import StoreContext from "./StoreContext";
import withNodesToProps from "./withNodesToProps";

const ignoreProps = ["children"];
let State;

export const withStore = WrappedComponent => {
  const WithStore = props => (
    <StoreContext.Consumer>
      {from => <WrappedComponent {...props} from={props.from || from} />}
    </StoreContext.Consumer>
  );

  WithStore.displayName = "withStore";
  WithStore.WrappedComponent = WrappedComponent;

  return hoistNonReactStatic(WithStore, WrappedComponent);
};

class InnerState extends Component {
  static Debug = false;

  static Verbose = false;

  constructor(props) {
    super(props);
    this.setup(true);
  }

  componentDidMount() {
    // call onMount handler
    const { onMount, from, path, default: def } = this.props;

    if (!onMount) return;

    let state = connect(from);
    if (path) state = getIn(state, path, def);

    if (onMount.length === 2) onMount(this.props, state);
    else onMount(state);
  }

  shouldComponentUpdate(nextProps) {
    if (this.prevStoreState !== this.storeState) {
      if (this.debug) this.log("shouldComponentUpdate", "store state change");
      return true;
    }

    // ignore children changes -- if you need a reset use render property
    // instead of children
    if (!shallowEqual(this.props, nextProps, ignoreProps)) {
      if (this.debug) this.log("shouldComponentUpdate", "props change");
      this.setup();
      return true;
    }

    if (this.debug && this.verbose)
      this.log("-shouldComponentUpdate", "skip render");

    return false;
  }

  componentWillUnmount() {
    this.reset();

    // call onUnmount handler
    const { onUnmount, from, path, default: def } = this.props;

    if (!onUnmount) return;

    let state = connect(from);
    if (path) state = getIn(state, path, def);

    if (onUnmount.length === 2) onUnmount(this.props, state);
    else onUnmount(state);
  }

  // setup this component
  setup = first => {
    // setup debugging
    this.setupDebug();
    this.reset();

    const { from, throttle: t, debounce: d, register: reg } = this.props;

    // setup the initial store state, registering state if requested
    if (first && reg) {
      if (this.debug) this.log("setup", "registering state", reg);
      let state = register(reg, from);
      if (state === undefined)
        state = connect(from, ReactDOM.unstable_batchedUpdates);
      this.trapSelect(state);
    } else this.trapSelect(connect(from, ReactDOM.unstable_batchedUpdates));

    // apply throttle or debounce
    if (t) this.onStateChange = throttle(t, this.origOnStateChange);
    else if (d) this.onStateChange = debounce(d, this.origOnStateChange);
    else this.onStateChange = this.origOnStateChange;
  };

  // setup debugging
  setupDebug = () => {
    // set component debug flag
    const { debug, verbose, id, path, from } = this.props;

    this.debug = State.Debug || debug;
    if (!this.debug) return;

    // set component verbose flag
    this.verbose = State.Verbose || verbose;

    // setup log prefix and logger
    let loggerId = id ? `State:${id}` : "State";
    if (path) loggerId += `:${path}`;
    if (from) loggerId = `${from}.${loggerId}`;

    // create logger function
    this.log = logger(loggerId);
  };

  // reset internal state
  reset = () => {
    this.watch = undefined;
    this.prevStoreState = undefined;
    this.prevFinalState = undefined;

    // cancel any pending debounced/throttled state changes
    if (this.onStateChange && this.onStateChange.cancel)
      this.onStateChange.cancel();

    // unsubscribe from state changes
    if (this.unsubscribe) this.unsubscribe();
  };

  // process a state change
  origOnStateChange = (state, changes) => {
    if (!changes || changes.length === 0) {
      if (this.debug) this.log("`no changes`");
      return;
    }

    // log if these changes have been delayed via debounce or throttle
    const { debounce: d, throttle: t } = this.props;

    if (this.debug && (d || t))
      this.log((d && "debounced") || "throttled", "changes", changes);

    this.trapSelect(state);
    this.forceUpdate();
  };

  // handle a state change notification
  stateSubscriber = (state, changes) => {
    if (this.debug) this.log("update", changes, "watch:", this.watch);

    this.onStateChange(state, changes);
  };

  // query state, potentially watching for fields accessed
  trapStateQuery = (state, select) => {
    const { from, path, default: def, constant, strict } = this.props;

    // handle constant
    if (constant) {
      const finalState = select(getIn(state, path, def));

      // log the constant query
      if (this.debug) this.log("get", "constant");

      return finalState;
    }

    // handle watched fields index is not yet available or strict is on
    if (this.watch === undefined || strict) {
      // execute a query, watching the fields accessed
      const [finalState, watch] = watchGet(state, select, path, def, from);
      this.watch = watch;

      // log the newly created watch fields index
      if (this.debug) this.log("watch", this.watch);

      // subscribe to changes specified by the watch index
      this.unsubscribe = subscribe(this.stateSubscriber, this.watch, from);

      return finalState;
    }

    // just execute the query, no need to track fields accessed because:
    // - not a constant,
    // - not a strict,
    // - has watched field index
    const finalState = select(getIn(state, path, def));
    return finalState;
  };

  // trap a select query, tracking fields accessed
  trapSelect = state => {
    // return original state when there's no select because fields are tracked
    // in the render function
    const { select } = this.props;
    this.storeState = select ? this.trapStateQuery(state, select) : state;
  };

  // trap a render function, tracking fields accessed
  trapRender = renderFn => {
    // strict equality means we have rendered for this state already
    if (this.prevStoreState === this.storeState) {
      // log a skip message because this update could probably be fixed upstream
      if (this.debug && this.verbose)
        this.log("-skip render", "prevState === nextState");

      return this.prevFinalState;
    }

    // constants only execute render function once
    const { constant } = this.props;

    if (constant && this.prevFinalState) {
      // log a skip message because this update could probably be fixed upstream
      if (this.debug && this.verbose) this.log("-skip render", "constant");

      return this.prevFinalState;
    }

    // execute the render function, trapping fields accessed
    const finalState = this.trapStateQuery(this.storeState, renderFn);

    // log the render
    if (this.debug)
      if (constant) this.log("render", "fn", "constant");
      else this.log("render", "fn");

    // cache the store state and the final state (result of render)
    this.prevStoreState = this.storeState;
    this.prevFinalState = finalState;
    return finalState;
  };

  render() {
    const { children, render, select, constant } = this.props;

    // select children or render as the render function
    const renderFn = children || render;

    // handle select query
    if (select) {
      // log a select render
      if (this.debug)
        if (constant) this.log("render", "select", "constant");
        else this.log("render", "select");

      // update prev store state
      this.prevStoreState = this.storeState;

      // no render function renders select result
      if (!renderFn) return this.storeState;

      // execute the render function with the selected state
      return renderFn(this.storeState);
    }

    // no render function renders nothing
    if (!renderFn || typeof renderFn !== "function") return null;

    // no select specified, trap the render function and return the result
    return this.trapRender(renderFn);
  }
}

State = withStore(withNodesToProps(InnerState));

InnerState.contextType = StoreContext;
InnerState.composes = [
  "register",
  "render",
  "onMount",
  "onUnmount",
  "select",
  "path",
  "default",
  "throttle",
  "ms",
  "constant",
  "strict",
  "debounce",
  "ms",
  "debug",
  "verbose",
  "id"
];

State.defaultProps = {
  path: "",
  from: undefined,
  select: undefined,
  constant: false,
  strict: false,
  default: undefined,
  throttle: undefined,
  debounce: undefined,
  debug: false,
  verbose: false,
  id: undefined,
  onMount: undefined,
  onUnmount: undefined,
  render: undefined
};

const StateComponent = State;

export default StateComponent;
