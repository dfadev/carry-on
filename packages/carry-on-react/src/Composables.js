import React from "react";
import State from "./State";

export function Register({ children, state, ...props }) {
  return <State register={{ state: state || children }} {...props} />;
}
Register.prop = "register";
Register.val = "children";
Register.transform = state => ({ state });
export function Middleware({ children, ...props }) {
  return <State register={{ middleware: children }} {...props} />;
}
Middleware.prop = "register";
Middleware.val = "children";
Middleware.transform = middleware => ({ middleware });
export const Render = () => null;
Render.prop = "render";
export const OnMount = () => null;
OnMount.prop = "onMount";
export const OnUnmount = () => null;
OnUnmount.prop = "onUnmount";
export const Select = () => null;
Select.prop = "select";
export const Path = () => null;
Path.prop = "path";
export const Default = () => null;
Default.prop = "default";
export const Throttle = () => null;
Throttle.prop = "throttle";
Throttle.val = "ms";
Throttle.default = 0;
export const Constant = () => null;
Constant.prop = "constant";
Constant.default = true;
export const Strict = () => null;
Strict.prop = "strict";
Strict.default = true;
export const Debounce = () => null;
Debounce.prop = "debounce";
Debounce.val = "ms";
Debounce.default = 0;
export const Debug = () => null;
Debug.prop = "debug";
Debug.val = "enabled";
Debug.default = true;
export const Verbose = () => null;
Verbose.prop = "verbose";
Verbose.val = "enabled";
Verbose.default = true;
export const Id = () => null;
Id.prop = "id";
