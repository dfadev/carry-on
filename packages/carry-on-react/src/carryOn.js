/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import State from "./State";

export default (renderFn, stateProps) => {
  if (typeof renderFn === "object") {
    const actualStateProps = renderFn;
    renderFn = stateProps;
    stateProps = actualStateProps;
  }

  const CarryOnComponent = ({ from, debug, verbose, onMount, onUnmount, ...props }) => {
    const finalProps = { ...stateProps };
    if (from !== undefined) finalProps.from = from;
    if (debug !== undefined) finalProps.debug = debug;
    if (verbose !== undefined) finalProps.verbose = verbose;
    if (onMount !== undefined) finalProps.onMount = onMount;
    if (onUnmount !== undefined) finalProps.onUnmount = onUnmount;
    /* eslint-disable-next-line */
    if (props.id !== undefined) finalProps.id = props.id;

    return (
      <State {...finalProps}>
        {state =>
          renderFn.length === 1 ? renderFn(state) : renderFn(props, state)
        }
      </State>
    );
  };

  CarryOnComponent.displayName = (stateProps && stateProps.id) || "CarryOn";

  return CarryOnComponent;
};
