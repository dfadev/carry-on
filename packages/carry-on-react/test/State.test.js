import React from "react";
import { render } from "@testing-library/react";
import { withState, State } from "../src";
import { getStore, deleteStore, register } from "carry-on-store";

test("<State /> renders", () => {
  const { asFragment } = render(<State>{() => "ok"}</State>);
  expect(asFragment()).toMatchSnapshot();
  deleteStore();
});

test("withState renders", () => {
  const Comp = withState()(props => "ok");
  const { asFragment } = render(<Comp />);
  expect(asFragment()).toMatchSnapshot();
  deleteStore();
});

test("withState selection function returns non Object as state prop", () => {
  const select = state => "ok";
  const Comp = withState({ select })(props => JSON.stringify(props));
  const { asFragment } = render(<Comp prop1="ok" />);
  expect(asFragment()).toMatchSnapshot();
  deleteStore();
});

test("withState selection function returns array as state prop", () => {
  const select = (state, props) => [1, 2, 3];
  const Comp = withState({ select })(props => JSON.stringify(props));
  const { asFragment } = render(<Comp prop1="ok" />);
  expect(asFragment()).toMatchSnapshot();
  deleteStore();
});

test("withState path, from, and default can be a values", () => {
  const select = (state, props) => state;
  const from = undefined;
  const path = "key1";
  const def = "default Value";

  register({
    state: ({ set }) => ({
      key1: "val1",
      key2: "val2"
    })
  });

  const Comp = withState({ select, from, path, def })(props =>
    JSON.stringify(props)
  );

  const App = props => <Comp lookup="key1" />;

  const { asFragment } = render(<App />);

  expect(asFragment()).toMatchSnapshot();
  deleteStore();
});

test("withState asProp", () => {
  const Comp = withState({ asProp: "stuff" })(props => JSON.stringify(props));
  const { asFragment } = render(<Comp prop1="ok" />);
  expect(asFragment()).toMatchSnapshot();
  deleteStore();
});
