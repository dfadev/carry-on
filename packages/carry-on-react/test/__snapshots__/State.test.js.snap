// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<State /> renders 1`] = `
<DocumentFragment>
  ok
</DocumentFragment>
`;

exports[`withState path, from, and default can be a values 1`] = `
<DocumentFragment>
  {"lookup":"key1","state":"val1"}
</DocumentFragment>
`;

exports[`withState renders 1`] = `
<DocumentFragment>
  ok
</DocumentFragment>
`;

exports[`withState selection function returns array as state prop 1`] = `
<DocumentFragment>
  {"prop1":"ok","state":[1,2,3]}
</DocumentFragment>
`;

exports[`withState selection function returns non Object as state prop 1`] = `
<DocumentFragment>
  {"prop1":"ok","state":"ok"}
</DocumentFragment>
`;
